### PR TITLE
Support for DPoP tokens

### PR DESCRIPTION
**Pull request by martinmine**
This PR implements support for [Demonstrating Proof of Possession (DPoP) RFC 9449](https://datatracker.ietf.org/doc/html/rfc9449) in both the Authorization Code flow (Fhi.HelseId.Web) and incoming API calls (Fhi.HelseId.Api). 
The authorization code flow support (web) is based on the implementation in Duende (see https://github.com/DuendeSoftware/Duende.AccessTokenManagement/blob/b0f2ac089375cecaf7c2f3f83bb52af6ded0e54a/src/Duende.AccessTokenManagement.OpenIdConnect/ConfigureOpenIdConnectOptions.cs), while the API support is based on sample code from NHN (see https://github.com/NorskHelsenett/HelseID.Samples/tree/master/SampleApi).

DPoP is configured through two flags in appsettings (identical for both Web and Api):

```json
  "HelseIdWebKonfigurasjon": {
    ....
    "RequireDPoPTokens": true
  }
```

We can scrap this code and instead rely on code in AspNetCore when Microsoft adds support for DPoP in their authentication stack, see: https://github.com/dotnet/aspnetcore/issues/58016

Resolves https://github.com/FHIDev/fhi.helseid/issues/364

### CODE REVIEW

**Comment by martinmine** [comment]
I have also added a warning message that is logged on application startup in case DPoP is not enabled:
> [12:22:54 WRN] This web-API is not configured to require DPoP-tokens. This must be enabled in order to be compliant with the NHN Security profile. It can be enabled by setting the flag RequireDPoPTokens to true in appsettings.json under 'HelseIdApiKonfigurasjon'.

File: Fhi.HelseId.Api/DPoP/DPoPProofValidationData.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+using System.Security.Cryptography;
+using System.Text;
+using Microsoft.AspNetCore.Http;
+using Microsoft.IdentityModel.Tokens;
+
+namespace Fhi.HelseId.Api.DPoP;
+
+public class DPoPProofValidationData
+{
+    public DPoPProofValidationData(HttpRequest request, string proofToken, string accessToken, string? cnfClaimValueFromAccessToken)
```
Comment (yne-bv) [inline Fhi.HelseId.Api/DPoP/DPoPProofValidationData.cs:11]: 💭 
Kan bruke primary constructor og sette verdiene direkte på propertiene

File: Fhi.HelseId.Api/DPoP/DPoPProofValidator.cs [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        SecurityAlgorithms.RsaSsaPssSha512,
+
+        SecurityAlgorithms.EcdsaSha256,
+        SecurityAlgorithms.EcdsaSha384,
+        SecurityAlgorithms.EcdsaSha512
+    };
+
+    private readonly IReplayCache _replayCache;
+
+    public DPoPProofValidator(IReplayCache replayCache)
```
Comment (yne-bv) [inline Fhi.HelseId.Api/DPoP/DPoPProofValidator.cs:41]: Kan bruke primary constructor her.

File: Fhi.HelseId.Web/DPoP/BackchannelHandler.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    IDPoPTokenCreator tokenHelper)
+    : DelegatingHandler
+{
+    private const string DPoPHeader = "DPoP";
+    private const string DPoPNonceHeader = "DPoP-Nonce";
+
+    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+    {
+        var httpContext = httpContextAccessor.HttpContext;
+        var requestHasJktContext = httpContext!.Items.ContainsKey("request_has_jkt_context");
```
Comment (yne-bv) [inline Fhi.HelseId.Web/DPoP/BackchannelHandler.cs:21]: `request_has_jkt_context` er i klassen `DPoPContext`, kjekt hvis du bruker den her, og i testene

File: Fhi.HelseId.Web/DPoP/DPoPContext.cs [hunk 4]
```diff
@@ -0,0 +1,6 @@
+﻿namespace Fhi.HelseId.Web.DPoP;
+
+public class DPoPContext
+{
+    public const string ContextKey = "request_has_jkt_context";
```
Comment (yne-bv) [inline Fhi.HelseId.Web/DPoP/DPoPContext.cs:5]: Er dette navnekonvensjonen?

Comment (martinmine) [reply Fhi.HelseId.Web/DPoP/DPoPContext.cs:5]: Det har ikke noe å si ettersom det kun er en verdi vi bruker internt i tjenesten for å identifisere om et HTTP-kall er gjort i kontekst av DPoP i HttpContexten

File: Fhi.HelseId.Api/DPoP/DPoPExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+public static class DPoPExtensions
+{
+    private const string DPoPAuthorizationSchema = AuthorizationScheme.DPoP + " ";
+
+    public static bool TryGetDPoPAccessToken(this HttpRequest request, out string token)
+    {
+        token = "";
+        var authorization = request.Headers.Authorization.SingleOrDefault() ?? "";
+        if (string.IsNullOrEmpty(authorization) || !authorization.StartsWith(DPoPAuthorizationSchema))
```
Comment (OsirisTerje) [inline Fhi.HelseId.Api/DPoP/DPoPExtensions.cs:16]: Skal være ikke case-sens.

File: Fhi.HelseId.Api/DPoP/DPoPExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
+using System.Linq;
+using Fhi.HelseId.Common.DPoP;
+using Microsoft.AspNetCore.Http;
+using Microsoft.IdentityModel.Tokens;
+
+namespace Fhi.HelseId.Api.DPoP;
+
```
Comment (OsirisTerje) [inline Fhi.HelseId.Api/DPoP/DPoPExtensions.cs:7]: Kommentar med link til hvor dette er hentet fra

File: Fhi.HelseId.Api/DPoP/DPoPExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            return false;
+        }
+        token = authorization.Substring(DPoPAuthorizationSchema.Length);
+        return true;
+    }
+
+    public static bool TryGetDPoPProof(this HttpRequest request, out string dPopProof)
+    {
+        dPopProof = request.Headers[DPoPHttpHeaders.ProofHeaderName].SingleOrDefault() ?? "";
```
Comment (OsirisTerje) [inline Fhi.HelseId.Api/DPoP/DPoPExtensions.cs:26]: Må sjekke at det kun er -1-.

File: Fhi.HelseId.Api/DPoP/JwtBearerDPoPTokenHandler.cs [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            {
+                tokenValidatedContext.Fail("Request has no DPoP token, which is required");
+            }
+
+            return;
+        }
+
+        if (!request.TryGetDPoPProof(out var dPopProof))
+        {
+            tokenValidatedContext.Fail("Missing DPoP proof");
```
Comment (OsirisTerje) [inline Fhi.HelseId.Api/DPoP/JwtBearerDPoPTokenHandler.cs:51]: Must be exactly one DPop Header

**Review by OsirisTerje — CHANGES_REQUESTED** [review]

**Review by OsirisTerje — APPROVED** [review]

**Comment by Rudde** [comment]
Nice

**Pull request merged**